sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/higherOrderFunctions.scala

```diff
@@ -369,6 +383,9 @@ case class ArrayFilter(
     var i = 0
     while (i < arr.numElements) {
       elementVar.value.set(arr.get(i, elementVar.dataType))
+      if (indexVar.isDefined) {
```

Comment:
Can you avoid this per-row check? The current code causes unnecessary runtime overheads.

Reply:
@maropu do you have a suggestion about how to do this without implementing codegen? I tried rewriting the logic like so:

```scala
  @transient private lazy val evalFn: (InternalRow, Any) => Any = indexVar match {
    case None => (inputRow, argumentValue) =>
      val arr = argumentValue.asInstanceOf[ArrayData]
      val f = functionForEval
      val buffer = new mutable.ArrayBuffer[Any](arr.numElements)
      var i = 0
      while (i < arr.numElements) {
        elementVar.value.set(arr.get(i, elementVar.dataType))
        if (f.eval(inputRow).asInstanceOf[Boolean]) {
          buffer += elementVar.value.get
        }
        i += 1
      }
      new GenericArrayData(buffer)

    case Some(expr) => (inputRow, argumentValue) =>
      val arr = argumentValue.asInstanceOf[ArrayData]
      val f = functionForEval
      val buffer = new mutable.ArrayBuffer[Any](arr.numElements)
      var i = 0
      while (i < arr.numElements) {
        elementVar.value.set(arr.get(i, elementVar.dataType))
        expr.value.set(i)
        if (f.eval(inputRow).asInstanceOf[Boolean]) {
          buffer += elementVar.value.get
        }
        i += 1
      }
      new GenericArrayData(buffer)

  }


  override def nullSafeEval(inputRow: InternalRow, argumentValue: Any): Any = {
    evalFn(inputRow, argumentValue)
  }
```

But from some hacky microbenchmarking this doesn't seem to be meaningfully faster and if anything is marginally slower.

Reply:
This is the benchmark code I was using:

Reply:
@maropu @henrydavidge The best performing way to avoid the per-row check in a non-codegen setting is to introduce a new expression type, say `ArrayFilterWithIndex`.

Reply:
I thought code like this;
```

  @transient lazy val (elementVar, mayFillIndex) = function match {
    case LambdaFunction(_, Seq(elemVar: NamedLambdaVariable), _) =>
      (elemVar, (_: Int) => {})
    case LambdaFunction(_, Seq(elemVar: NamedLambdaVariable, idxVar: NamedLambdaVariable), _) =>
      (elemVar, (i: Int) => idxVar.value.set(i))
  }

  override def nullSafeEval(inputRow: InternalRow, argumentValue: Any): Any = {
    val arr = argumentValue.asInstanceOf[ArrayData]
    val f = functionForEval
    val buffer = new mutable.ArrayBuffer[Any](arr.numElements)
    var i = 0
    while (i < arr.numElements) {
      elementVar.value.set(arr.get(i, elementVar.dataType))
      mayFillIndex(i)
      if (f.eval(inputRow).asInstanceOf[Boolean]) {
        buffer += elementVar.value.get
      }
      i += 1
    }
    new GenericArrayData(buffer)
  }
```

Reply:
Ok, tried that as well. It doesn't seem to be significantly different from the others.

Reply:
Yea, if no big difference, I like the similar handling with the others, e.g., https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala#L555

Reply:
I think this is good enough to go.
How about merging this for now, and addressing it in a separate PR?
`transform` is doing the same way, so I think we should do the same thing if needed, maybe at the same time.

Reply:
+1 for this is ready to go for now and we can address the optimization separately.

Side-comment on the version that @maropu gave:
The version that @henrydavidge gave would technically have less overhead:
```
// lambda invocation overhead outside of loop
for each element in array
  do specialized filter action
```
whereas the version that @maropu gave would be:
```
// shared loop between the two versions
for each element in array
  // lambda invocation overhead per element
  invoke mayFillIndex lambda
```

With @maropu 's version, let's assume that we're running on the HotSpot JVM and both the with-index and without-index paths have been used, then the best HotSpot JIT could have done is a profile-guided bimorphic devirtualization on that lambda call site, which will look like the following after devirtualization+inlining:
```
local_mayFillIndex = this.mayFillIndex
for each element in array
  klazz = local_mayFillIndex.klass
  if (klazz == lambda_klass_1) {
    // no-op
  } else if (klazz == lambda_klass_2) {
    idxVar.value.set(i)
  }
}
```
The point is that this JIT-optimized version is actually a degenerated version of Henry's hand-written inline per-row check version, so I wouldn't want to go down this route.

Reply:
Thanks, kris! That explanation's very helpful to me.